libsel4vm/src/arch/arm/vgic/vgic.c

```diff
@@ -512,11 +512,9 @@ static int vgic_vcpu_inject_irq(struct vgic_dist_device *d, vm_vcpu_t *inject_vc
     if (!err) {
         /* Shadow */
         vgic->irq[inject_vcpu->vcpu_id][i] = irq;
-        return err;
-    } else {
-        /* Add to overflow list */
-        return vgic_add_overflow(vgic, irq, inject_vcpu);
     }
+
```

Comment:
Putting an `assert(!vgic_irq_queued(vgic, vcpu, irq))` might be too expensive, but this function should get a comment that that at this stage the callmer must have ensured the IRQ is not queued and thus there is no need to clean the queue here. I have been working on this code also and more comment everywhere are really helpful.

Reply:
In this case, the comment you suggested would not be entirely correct. Sure, if it is called from vgic_dist_set_pending_irq(), it is true, but when called by the vGIC maintenance interrupt, the IRQ is in the overflow queue; otherwise we wouldn't be trying to reinject that in the first place. Instead, I rewrote code and tried to made it clear that we are only trying to load the IRQ into one of the four hardware-backed list registers and that's all.